### PR TITLE
Dense CuPy cython level and inheritance to data.Data

### PR DESCRIPTION
**Pull request by MrRobot2211**
I tried to create the new Dense CuPy class at cython level and to  inherit from `data.Data` 
In this case the project fails to build. I suspect there is a problem when lnking to `dev.major`. I tried a workaround to set the requirement on the `toml` file, and as it is not released to explicit the github url for the download and it still failed.

### CODE REVIEW

File: src/qutip_cupy/cupy_dense.pyx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import cupy as cp 
+import qutip
+
+
+from qutip.core.data cimport base
+
+
+
+cdef class Dense(data.Data):
```
Comment (Ericgig) [inline src/qutip_cupy/cupy_dense.pyx:17]: You import only `base` from `qutip` so `data.Data` does not exist.

Comment (MrRobot2211) [reply src/qutip_cupy/cupy_dense.pyx:17]: Sorry about that. I have been trying many ways to import it. Still when changing to `base.Data` it is not finding the package, I changed the requirements as suggested by @jakelishman  on #7  but there seems to be still a linking problem.

**Comment by jakelishman** [comment]
You have a few problems here:

1. as it stands, this probably won't build locally because `np` isn't defined in your `.pyx` file, but you use it.
2. you depend on `qutip` at build time, but you haven't specified it in `pyproject.toml`.  Also, you don't depend on `scipy` at build, but you _have_ specified it.
3. you'll need to pass the `--no-build-isolation` flag to `pip install` to get it to use an existing QuTiP installation, otherwise it'll try and source a clean one from PyPI to use during the build.

I can also see you've pulled the `setup.py` from QuTiP - unfortunately at the time you pulled it, `dev.major` was using an unfortunately out-dated process.  It still works, but it's rather more manual than it needs to be (you have to add your Cython files one-by-one).  If you look at the newer version (which is _now_ on `dev.major`), it's a much better template.

I can't push changes to this PR, but here's a patch which would fix the Cython issues:
```
From 8b6ee19c179fcc62354b517caa11ed2539eb8c06 Mon Sep 17 00:00:00 2001
From: Jake Lishman <jake@binhbar.com>
Date: Wed, 16 Jun 2021 23:48:53 +0100
Subject: [PATCH] Force CI pip builds to use existing QuTiP

pip will try to install a new version of QuTiP as part of its build
isolation process, but this will fail as long as the versions we need
aren't accessible on PyPI.
---
 .github/workflows/docs.yml  | 5 ++++-
 .github/workflows/tests.yml | 6 +++++-
 pyproject.toml              | 1 +
 3 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/.github/workflows/docs.yml b/.github/workflows/docs.yml
index a9ae32c..4e22ceb 100644
--- a/.github/workflows/docs.yml
+++ b/.github/workflows/docs.yml
@@ -26,8 +26,11 @@ jobs:
           python -c 'import qutip; qutip.about()'
 
       - name: Install qutip-cupy from GitHub
+        # The --no-build-isolation flag in the pip install command is necessary
+        # to force pip to use the version of QuTiP we just installed, rather
+        # than trying to find it on PyPI.
         run: |
-          python -mpip install .[full]
+          python -mpip install --no-build-isolation .[full]
           python -c 'import qutip_cupy; print(qutip_cupy.__version__)'
 
       - name: Build documentation
diff --git a/.github/workflows/tests.yml b/.github/workflows/tests.yml
index 67a5604..534043c 100644
--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -42,8 +42,12 @@ jobs:
       - name: Install qutip-cupy and dependencies
         # Install in editable mode so Coveralls detects the tree structure
         # relative to the git repository root as well.
+        #
+        # The --no-build-isolation flag is necessary to get pip to use the
+        # version of QuTiP we already installed.  Otherwise, it will try to
+        # search its index for a prerelease that does not yet exist.
         run: |
-            python -mpip install -e .[full]
+            python -mpip install --no-build-isolation -e .[full]
             python -mpip install pytest-cov coveralls
 
       - name: Package information
diff --git a/pyproject.toml b/pyproject.toml
index 5e60927..4622002 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -3,6 +3,7 @@ requires = [
     "setuptools>=36.6",
     "packaging",
     "wheel",
+    "qutip>=5.0.0.dev0",
     "cython>=0.29.20",
     "numpy","scipy"
 ]
-- 
2.32.0
```

**Comment by MrRobot2211** [comment]
Thank you @jakelishman the `--no-build-isolation` does the trick, now I can  include QuTiP dev.major in the `toml`  without throwing an error

**Comment by MrRobot2211** [comment]
Okey it looks like by setting the `--no-build-isolation` we have to manually install all dependences . I probably can just loop and pip install through the toml, the installation requirements and the setup requirements. Or is there a more elegant option?

I will try to see if we can get the CUDA emulator running so that CuPy  can get installed here.

**Review by hodgestar — COMMENTED** [review]
As described in #4, I think we can avoid and should avoiding working with Cython initially -- we have enough complexity to deal with in CuPy and CUDA themselves, and since CuPy should already provide efficient Python methods for the common operations we want to perform, we likely won't benefit initially from writing things in Cython ourselves.

I think we should close this PR and focus on #4 instead.

**Comment by MrRobot2211** [comment]
Thanks you all, closing this one then.

**Pull request closed without merging**